File: app/src/main/cpp/WavetableSynth.cpp
//
// Created by ched54 on 3/19/2024.
//
#include "include/Log.h"
#include "WavetableSynth.h"
#include "OboeAudioPlayer.h"
#include "WavetableOsc.h"


namespace wavetablesynth {

    WavetableSynth::WavetableSynth() : _oscillator{std::make_shared<A4Osc>(sampleRate)},
                                       _audioPlayer{std::make_unique<OboeAudioPlayer>(_oscillator,
                                                                                      sampleRate)} {
        LOGD("WavetableSynth::WavetableSynth()");

    }
    WavetableSynth::~WavetableSynth() = default;


    void WavetableSynth::play() {
        LOGD("WavetableSynth::play()");
        const auto result = _audioPlayer->play();
        if (result == 0){
            _isPlaying = true;
        } else {
            LOGD("Failed to play audio");
        }
    }

    void WavetableSynth::stop() {
        LOGD("WavetableSynth::stop()");
        _audioPlayer->stop();
        _isPlaying = false;
    }

    bool WavetableSynth::isPlaying() const{
        LOGD("WavetableSynth::isPlaying()");
        return _isPlaying;
    }

    void WavetableSynth::setFrequency(float frequency) {
        LOGD("WavetableSynth::setFrequency() called with %.2f", frequency);
    }

    void WavetableSynth::setVolume(float volume) {
        LOGD("WavetableSynth::setVolume() called with %.2f", volume);
    }

    void WavetableSynth::setWavetable(wavetablesynth::Wavetable wavetable) {
        LOGD("WavetableSynth::setWaveform() called with %d", static_cast<int>(wavetable));
    }



}

File: app/src/main/java/me/typosbro/synth/WavetableSynth.kt
package me.typosbro.synth

import androidx.annotation.StringRes

enum class Wavetable {
    SINE {
        override fun toResourceString(): Int {
            return R.string.sine
        }
    },
    TRIANGLE {
        override fun toResourceString(): Int {
            return R.string.triangle
        }
    },
    SQUARE {
        override fun toResourceString(): Int {
            return R.string.square
        }
    },
    SAWTOOTH {
        override fun toResourceString(): Int {
            return R.string.sawtooth
        }
    };

    @StringRes
    abstract fun toResourceString(): Int
}

interface WavetableSynth {
    suspend fun play()
    suspend fun stop()
    suspend fun isPlaying(): Boolean
    suspend fun setFrequency(frequency: Float)
    suspend fun setVolume(volume: Float)
    suspend fun setWavetable(wavetable: Wavetable)

}